Iao Seng Sio ICS434

### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.



### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame`s or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitely asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external products will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora or functionality available through the `pandas` package.

---

In [2]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [3]:
## WRITE YOUR CODE HERE 
tax_data = pd.read_csv('data/tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [6]:
## WRITE YOUR CODE HERE 
tax_data.head(8)

STATEFIPS STATE  ZIPCODE  AGI_STUB        N1     MARS1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2    ...        N10300     A10300   N85530   A85530  \
0  455560.0  1356760.0    ...      373410.0   328469.0      0.0      0.0   
1  275920.0  1010990.0    ...      395880.0   965011.0      0.0      0.0   
2  155100.0   583910.0    ...      251490.0  1333418.0      0.0      0.0   
3   99950.0   423990.0    ...      165320.0  1414283.0      0.0      0.0   
4  126860.0   589490.0    ...      212000.0  3820152.0    420.0    168.0   
5   41410.0   160530.0    ...       55300.0  6027793.0  22090.0  39519.0   
6     700.0     2160.0    ...         690.0      610.0      0.0      0.0   
7     610.0     2540.0    ...        1140.0     3019.0      0.0      0.0   

    N85300   A85300   N11901    A11901    N11902     A11902  
0      0.0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0      0.0  45460.0  128823.0  116440.0   377177.0  
4     60.0     31.0  83330.0  421004.0  121570.0   483682.0  
5  27550.0  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0      0.0    120.0      94.0    1290.0     2792.0  
7      0.0      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [5]:
## WRITE YOUR CODE HERE 
tax_data.columns = tax_data.columns.str.upper()

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
## WRITE YOUR CODE HERE 
tax_data.shape[0]

166698

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [8]:
## WRITE YOUR CODE HERE 
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [9]:
## WRITE YOUR CODE HERE 
tax_data.columns.get_loc('N10300')

81

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zipcodes in TX, 9,238 in NY, etc.


In [10]:
## WRITE YOUR CODE HERE 
unique_zipcode = tax_data.groupby('STATE')['ZIPCODE'].nunique().sort_values(ascending=False)
unique_zipcode

STATE
TX    1619
NY    1540
CA    1485
PA    1368
IL    1231
OH     998
FL     920
MI     892
MO     890
IA     827
VA     793
MN     792
NC     725
WI     714
IN     674
GA     667
KY     656
KS     598
TN     590
AL     577
OK     547
NJ     547
WV     512
WA     497
AR     490
NE     486
MA     482
LA     453
MD     405
CO     395
SC     376
ME     370
MS     370
OR     356
AZ     295
ND     287
SD     287
CT     264
VT     238
NH     232
MT     225
ID     213
NM     209
UT     186
NV     128
WY     109
RI      71
HI      60
DE      57
AK      56
DC      24
Name: ZIPCODE, dtype: int64

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [11]:
## WRITE YOUR CODE HERE 
unique_zipcode.index.get_loc('HI')

47

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [12]:
## WRITE YOUR CODE HERE 
nb_invalid_zip = tax_data[(tax_data['ZIPCODE'] == 0)].shape[0]

* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [13]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [14]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip = tax_data[(tax_data['ZIPCODE'] != 0)]

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [15]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [16]:
## WRITE YOUR CODE HERE 
nb_missing_values = tax_data_valid_zip.shape[0] - tax_data_valid_zip.dropna().shape[0]

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [17]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [18]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [19]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * Recall that you can use the `apply` function to transform the values of a column
  
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`
* The resulting `Series` should look like the following( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [20]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip_cleaned = tax_data_valid_zip_cleaned.astype({'ZIPCODE': 'int64'})
zip_rev_all = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip, percentile=0.65).sort_values(ascending=False)

- What are the three zip codes with the most significant 65th percentile value for income?

In [21]:
## WRITE YOUR CODE HERE 
zip_rev_all.head(3)

ZIPCODE
33109    3954.114286
33480    3413.301538
94301    3109.443711
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [22]:
## WRITE YOUR CODE HERE 
medicaid_data = pd.read_csv('data/medicaid_data.csv')


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [23]:
## WRITE YOUR CODE HERE 
medicaid_data.columns = medicaid_data.columns.str.upper()


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [24]:
## WRITE YOUR CODE HERE 
medicaid_data.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               LOCATION  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)

* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [25]:
## WRITE YOUR CODE HERE 
(medicaid_data.groupby('STATE')['LOCATION'].nunique() > 1).any()

False



* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [26]:
## WRITE YOUR CODE HERE 
medicaid_enrollment = pd.read_csv('data/medicaid_enrollment.tsv', '\t')

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [27]:
## WRITE YOUR CODE HERE 
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [28]:
## WRITE YOUR CODE HERE 
medicaid_enrollment['TOTAL MEDICAID ENROLLEES'] = pd.to_numeric(medicaid_enrollment['TOTAL MEDICAID ENROLLEES'].str.replace(' ', '').str.replace(',', ''), errors='coerce')
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [29]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned.astype({'TOTAL MEDICAID ENROLLEES': 'int64'})
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtype

dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [30]:
# reads all the tables at the given url
# Header=0 instructs pandas to use line 0 as the header (columns)
import requests

url = "https://www.50states.com/abbreviations.htm"
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}
r = requests.get(url, headers=header)
Codes_abbreviations = pd.read_html(r.text, header=0)[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US State:', 'Abbreviation:'] to ['US STATE', 'ABBREVIATION']

  * You can hard code this operation


In [32]:
## WRITE YOUR CODE HERE 
Codes_abbreviations.columns = ['US STATE', 'ABBREVIATION', 'STANDARD ABBREVIATION']

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online. See for instance:

`https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`

* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [30]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, left_on='STATE', right_on='US STATE')
medicaid_enrollment_cleaned_with_zip.drop('US STATE', axis=1, inplace=True)

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [31]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip.drop('STATE', axis=1, inplace=True)

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [32]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments = medicaid_data.merge(medicaid_enrollment_cleaned_with_zip, left_on='STATE', right_on='ABBREVIATION')
medicaid_data_w_enrollments.drop('ABBREVIATION', axis=1, inplace=True)

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [33]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments = medicaid_data_w_enrollments[medicaid_data_w_enrollments[['STATE', 'PRODUCT NAME']].notna().all(axis=1)]

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [34]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments.set_index(['STATE', 'NDC'])
medicaid_data_w_enrollments_hierarch

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    55150023930             MCOU   Dexamethas            33.000   
      456060010               FFSU   TEFLARO 60           224.000   
      115692201               MCOU   DIVALPROEX          1626.000   
      54455025                FFSU    METHOTREX           450.000   
      50458014090             MCOU     INVOKANA          3850.000   
      68682036990             MCOU    Diltiazem          3911.000   
      66582031131             MCOU   VYTORIN 10          1050.000   
      904585260               FFSU   CETIRIZINE           944.000   
      378037301               FFSU   HYDROXYCHL          2274.000   
      11980002205             MCOU   BLEPHAMIDE            65.000   
      781150601               MCOU   ATENOLOL 5          7040.000   
      68180040203             MCOU    CEFPROZIL         41455.000   
      781104601               FFSU   PERPHENAZI           248.000   
      45802004201             MCOU    Clindesse           133.400   
      31722053812             FFSU   LEVETIRACE         13647.000   
      603014921               FFSU    Doc-Q-Lax         36813.000   
      29300022013             MCOU   MONTELUKAS          1706.000   
      536108794               FFSU   NATURES TE           300.000   
      173077602               MCOU   LAMICTAL O          5370.000   
      63481081660             MCOU   OPANA ER 2          1447.000   
      64125090505             FFSU    LORAZEPAM          1193.000   
      50242010040             FFSU    PULMOZYME          5700.000   
      68382003305             FFSU   Divalproex          1822.000   
      591278886               MCOU   TIZANIDINE          1245.000   
      93510505                MCOU   OLANZAPINE          2821.000   
      59762015701             MCOU   ATORVASTAS          4751.000   
      172524160               MCOU   ANAGRELIDE          2254.000   
      67405060203             FFSU   Fluconazol            73.000   
      187540004               MCOU   JUBLIA 4ml            96.000   
      64380075906             MCOU   ACARBOSE T          1995.000   
...                            ...          ...               ...   
NJ    703367501               MCOU   METHOTREXA          1218.040   
      93823898                MCOU   IRBESARTAN          1706.000   
      2771227                 MCOU   HUMALOG KW           480.000   
      62856058252             MCOU   BANZEL 200          9040.000   
      68645054959             MCOU    Metformin          5144.000   
      378262510               MCOU   AMITRIPTYL         41031.000   
      47781029703             MCOU   NITROGLYCE          2288.000   
      68001026303             MCOU   BPL Metron          3988.000   
      57664022388             MCOU    OXYCODONE         26269.000   
      641605710               MCOU    Midazolam            57.000   
      591321472               MCOU   FENTANYL P           553.000   
      43199004030             MCOU    Lidocaine         70906.740   
      10702001706             MCOU   CHLORPHENI          2220.000   
      115169606               MCOU   Hydrocorti          2120.000   
      64679092403             MCOU    ENALAPRIL         13426.000   
      51079099120             MCOU   TRAMADOL H           109.000   
      378013701               MCOU   ALLOPURINO         18325.000   
      31722023910             MCOU   AMLODIPINE         62497.357   
      60505311108             MCOU   OLANZAPINE          1980.000   
      43547031509             MCOU   VALSARTAN/          8886.000   
      555904758               MCOU   ENPRESSE T          7616.000   
      378032710               MCOU   HALOPERIDO           878.000   
      13107005701             MCOU    OXYCODONE          7199.000   
      406838001               MCOU   MORPHINE S          4706.000   
      67457042100             MCOU   DEXAMETHAS           517.000   
      6332303146


* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [35]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch.loc[('PA', 61958180101)]

C:\Users\tonys\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: PerformanceWarning: indexing past lexsort depth may impact performance.
  


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232

#### Computing the normalized `UNITS REIMBURSED` per druc 

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the numebr of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

- Rather than work directly with the ratios, we are going to compute and work with thier logarithm (np.log2) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`


- Your final result should be a Series that look like the following ( `...` represents data that is not shown here):
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626

...


```


In [36]:
## WRITE YOUR CODE HERE 
def compute_normalized_units(df):
    return np.log2(df['UNITS REIMBURSED'].sum() / df['TOTAL MEDICAID ENROLLEES'].unique()[0])

medicaid_reimbursement_per_enrollee = medicaid_data_w_enrollments_hierarch.groupby(['STATE', 'NDC']).apply(compute_normalized_units)

- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [43]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc = medicaid_reimbursement_per_enrollee.unstack()

#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [61]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc.loc['HI'][(medicaid_norm_ndc.loc['HI'] == medicaid_norm_ndc.loc['HI'].max())]

NDC
43386009019    10.511181
Name: HI, dtype: float64

* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

In [70]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch.loc[('HI', 43386009019)]

C:\Users\tonys\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: PerformanceWarning: indexing past lexsort depth may impact performance.
  


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
HI    43386009019             MCOU   GAVILYTE-G          472000.0   
      43386009019             MCOU   GAVILYTE-G          376012.0   
      43386009019             MCOU   GAVILYTE-G       496104006.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
HI    43386009019                    118.0                  1754.68   
      43386009019                    105.0                  1462.60   
      43386009019                     63.0                   847.56   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
HI    43386009019                     1754.68                            0.00   
      43386009019                     1462.60                            0.00   
      43386009019                      839.61                            7.95   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
HI    43386009019  (21.1098, -157.5311)                    340513  
      43386009019  (21.1098, -157.5311)                    340513  
      43386009019  (21.1098, -157.5311)                    340513

## WRITE YOUR ANSWER HERE 
* It is used for bowel cleansing before colonoscopy.
* It may be a typo because other states have much lower numbers. The units reimbursed and the amount of money associated simply do not add up. There is no way that almost half a billion units have been reimbursed for less than a thousand dollars.

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [118]:
## WRITE YOUR CODE HERE 
medicaid_large_diff_ndc = medicaid_reimbursement_per_enrollee.groupby('NDC').max() >= medicaid_reimbursement_per_enrollee.groupby('NDC').nlargest(2).groupby('NDC').min() + 10
medicaid_large_diff_ndc[medicaid_large_diff_ndc].index

Int64Index([    9023303,   264761210,   338007304,   406036501,   407141233,
              409710002,   409767009,   409796709,   574082710,   591289749,
              641047721, 39822420006, 50458016601, 60793060002, 63323017530,
            68001028529, 68180062210],
           dtype='int64', name='NDC')

- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10